lib/spack/spack/environment.py

```diff
@@ -1305,31 +1305,30 @@ def _env_modifications_for_default_view(self, reverse=False):
         errors = []
         for _, root_spec in self.concretized_specs():
             if root_spec in self.default_view and root_spec.package.installed:
-                for spec in root_spec.traverse(deptype='run', root=True):
-                    if spec.name in visited:
-                        # It is expected that only one instance of the package
-                        # can be added to the environment - do not attempt to
-                        # add multiple.
-                        tty.debug(
-                            "Not adding {0} to shell modifications: "
-                            "this package has already been added".format(
-                                spec.format("{name}/{hash:7}")
-                            )
+                if root_spec.name in visited:
```

Comment:
Should I just also undo the `visited` part @scheibelp? It now only applies to root specs anyways.

Reply:
This is going to cause problems -- what if you have hdf5^mpich and openmpi in your environment, and openmpi is loaded first. Then the hdf5 dependency mpich will be loaded later than openmpi, and will shadow its environment variables.

I think we may be able to get around that problem by sorting the root specs by size -- if we always load bigger specs first, then they can't be "sneakily" pulling in something that will shadow a dependency (and then I think we could get rid of the `visited` list entirely.